File: algorithms/1_easy/S/00101_symmetric_tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def is_symmetric_helper(self, p, q):
        if p == None and q == None:
            return True

        if p == None or q == None or p.val != q.val:
            return False

        return self.is_symmetric_helper(p.left, q.right) and self. is_symmetric_helper(p.right, q.left)

    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
         return self.is_symmetric_helper(root.left, root.right)

File: algorithms/1_easy/S/00101_symmetric_tree/solution.c
static bool is_symmetric_helper(struct TreeNode* p, struct TreeNode* q)
{
    if (p == NULL && q == NULL) {
        return true;
    }

    if (p == NULL || q == NULL || p->val != q->val) {
        return false;
    }

    return is_symmetric_helper(p->left, q->right) && is_symmetric_helper(p->right, q->left);
}

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
bool isSymmetric(struct TreeNode* root)
{
    return is_symmetric_helper(root->left, root->right);
}
